### PR TITLE
Add OidcTenantConfig builder

### LINKED ISSUE #39185: Migrate OIDC-based extensions from config classes to `@ConfigMapping`
### Description

I've mentioned extensions migrating to `@ConfigMapping` along its own axis. Some of them try to solve a bug like Hibernate, but there are other objections like performance, efficiency and possibly new features (config classes are maintained if there is a major bug found, but not developed). Cons is that some config classes can be injected by integrating extensions or users, therefore they can be considered API - https://github.com/quarkusio/quarkus/pull/35246. However I don't believe that is a case of OIDC, how many extensions and users outside of Quarkus core extensions really needs to inject configuration classes as CDI beans, or use them in a build step?

I'd like to migrate OIDC config because I think impact will be absolutely minor. I don't know why should users inject OIDC config in general. Waiting probably solves nothing and migration can have positive performance impact (though migrating OIDC alone will hardly be measurable, it is one piece of the puzzle...).

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by michalvavrik**
- closes https://github.com/quarkusio/quarkus/issues/39185 (as this is final part, the deprecations and removals will take a long time)
- provide a builder for `OidcTenantConfig` class
- deprecate usage of `OidcTenantConfig` class fields and setters
- recommend using interface methods instead
- newly I also added deprecations on fields and inner classes of deprecated abstract classes `OidcCommonConfig` and `OidcClientCommonConfig` because my IDE doens't provide hints when these fields are used from the `OidcTenantConfig` instance and I want users to be warned; it is bit verbose though
- as I was already touching everything, I made some neglibible refactoring as well (`!optional.isPresent()` -> `optional.isEmpty`), not much though

These changes will allow us in the future in this order (with long breaks):
1. mark fields for removal
2. make sure users can't modify `OidcTenantConfig` directly, either:
  - make it only modifiable from OIDC runtime config
  - or better, we will make it immutable:
3. further:
  - fields will be final and private (maybe a record?)
  - users will use implemented interface methods to retrieve values
  - `OidcTenantConfig` will only be possible to build with the builder by users
  - drop `extends` to former config classes  and only leave `implements` for the interfaces
  - OidcTenantConfig class will be final
4. stop copying properties between builder and `OidcTenantConfig` as we can build the tenant config directly

This way, the API for users is in the `io.quarkus.oidc` package, more specifically:

- `io.quarkus.oidc.OidcTenantConfig`
- `io.quarkus.oidc.OidcTenantConfigBuilder`

### CODE REVIEW

**Comment by sberyozkin** [comment]
@michalvavrik Hi Michal, thanks, going forward, after this PR is merged, what happens when a new property has to be added, while the deprecated `OidcTenantConfig`  is still around ? I'm assuming that only `ConfigMapping` and the corresponding builder has to be updated, and if users who work with `TenantConfigProvider` would like to use a new property then they can only get it by using a builder, right ?

**Comment by michalvavrik** [comment]
> after this PR is merged, what happens when a new property has to be added, while the deprecated OidcTenantConfig is still around ?

1. I didn't deprecate this class. It is omnipresent including `RoutingContext` etc. It is unnecessary, we can keep it without issues. I think there are 2 ways it can go:
  - we all the field and parent class deprecations are removed, we can turn it into final class and remove private class `OidcTenantConfigImpl` from the builder. it will make things bit more efficient, we will also be able a lot of copying of properties etc. (more effective) [it would require moving some logic further to SR Config customizer)
  - we can keep it mutable but only from the Quarkus OIDC runtime package so that users only works with API, but we can keep changing things
2. If you add a new property, you are adding a new method to the interface. And `OidcTenantConfig` implements this interface, therefore you are guaranteed to be forced to update this config class appropriately. We can't forget otherwise compilation fails.

>  I'm assuming that only ConfigMapping and the corresponding builder has to be updated

yes; the former config class group `OidcTenantConfig` stiill implements this interface, but you will not add `public` field for this field. This way, you will force users to use the builder.

> and if users who work with TenantConfigProvider would like to use a new property then they can only get it by using a builder, right ?

they will need to use the builder if they want to create the `OidcTenantConfig` with this new property, but they can still do this:

```
OidcTenantConfig tenantConfig = OidcTenantConfig.builder().sergeysNewProperty(1).build();
if (1 == tenantConfig.sergeysNewProperty()) { // do stuff
```

This can be hard to understand, I can answer further questions to explain what I really mean.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: Michal, I'd not mind moving all sub-configmapping builders into the `.runtime` package and only keep a top level `OidcTenantConfigBuilder` builder, to allow for some longer term restructurings within those sub-mappings and keep `io.quarkus.oidc` API restricted only to the main `OidcTenantConfig`, does it sound it reasonable ? Otherwise the individual sub-builder may generate a bit of a noise in the API package

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: Well, ideally it would be `OidcTenantConfigBuilder` inner-class, but I was forced 3 largest inner classes outside to limit the `OidcTenantConfigBuilder` size.

> Otherwise the individual sub-builder may generate a bit of a noise in the API package

yeah, I agree about the noise. I can move them to the `runtime`.

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: > moving all sub-configmapping builders

All the `OidcTenantConfigBuilder` inner-classes? at the very least it will require to add getters for every built config group, like if I want to copy `Authentication`, I need to get it first. And I need to copy it if I want to get defaults (or previous values).

TL;DR; I'll change it if you are certain. It is not important IMO, so not an issue to do it either.

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: I want to get this PR done because it has a lot of merge-conflict potential, so I'll make these requested changes right now, won't wait for the confirmation.

Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: No, only independent classes like `AuthenticationConfigBuilder`, `LogoutConfigBuilder` and `TenantConfigBuilder`, that should be ok, right ?

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/LogoutConfigBuilder.java:1]: yes, that makes sense, thanks!

**Comment by sberyozkin** [comment]
@michalvavrik 

>>after this PR is merged, what happens when a new property has to be added, while the deprecated OidcTenantConfig is still around ?

>I didn't deprecate this class. It is omnipresent including RoutingContext etc. It is unnecessary, we can keep it without issues. I think there are 2 ways it can go:
   * when all the field and parent class deprecations are removed, we can turn it into final class and remove private class OidcTenantConfigImpl from the builder. it will make things bit more efficient, we will also be able a lot of copying of properties etc. (more effective) [it would require moving some logic further to SR Config customizer)
    * we can keep it mutable but only from the Quarkus OIDC runtime package so that users only works with API, but we can keep changing things
    * update: I just realized there is a third way - we could turn it into the interface and use it in the @ConfigMapping

I've looked at the code and I see `new io.quarkus.oidc.OidcTenantConfig()` is deprecated, the `io.quarkus.oidc.OidcTenantConfig` class itself should indeed not be deprecated as ultimately, it will be replaced by `io.quarkus.oidc.runtime.OidcTenantConfig`.

Should we mark though all the deprecated `io.quarkus.oidc.OidcTenantConfig` constructors, properties, for removal ?

>>and if users who work with TenantConfigProvider would like to use a new property then they can only get it by using a builder, right ?

>they will need to use the builder if they want to create the OidcTenantConfig with this new property, but they can still do this:
```
OidcTenantConfig tenantConfig = OidcTenantConfig.builder().sergeysNewProperty(1).build();
if (1 == tenantConfig.sergeysNewProperty()) { // do stuff
```
This can be hard to understand, I can answer further questions to explain what I really mean.

Sorry, I'm not sure what is the point that you are trying to highlight, I guess, once the `OidcTenantConfig` is ready, once can check its properties like you showed ?

**Comment by michalvavrik** [comment]
> I've looked at the code and I see new io.quarkus.oidc.OidcTenantConfig() is deprecated, the io.quarkus.oidc.OidcTenantConfig class itself should indeed not be deprecated as ultimately, it will be replaced by io.quarkus.oidc.runtime.OidcTenantConfig.

exactly, fields, setters and constructors are deprecated. I also deprecated getter in favor of the interface methods, but that one is not vital

> Should we mark though all the deprecated io.quarkus.oidc.OidcTenantConfig constructors, properties, for removal ?

I'd like that. I didn't mark them for removal because I tried to be conservative. So if you are fine with marking them for removal, we can do that. We could also wait for a release or two.

> Sorry, I'm not sure what is the point that you are trying to highlight, I guess, once the OidcTenantConfig is ready, one can check its properties like you showed ?

I tried to show that you can add new property without breaking things. Sorry if that was obvious.

**Comment by sberyozkin** [comment]
>OidcTenantConfig tenantConfig = OidcTenantConfig.builder().sergeysNewProperty(1).build();
if (1 == tenantConfig.sergeysNewProperty()) { // do stuff

Do you mean that works because it implements the interface ? It is fine, I don't really expect users do it, they will create `OidcTenantConfig` in dynamic resolvers only

**Comment by michalvavrik** [comment]
> > OidcTenantConfig tenantConfig = OidcTenantConfig.builder().sergeysNewProperty(1).build();
> > if (1 == tenantConfig.sergeysNewProperty()) { // do stuff
> 
> Do you mean that works because it implements the interface ? It is fine, I don't really expect users do it, they will create `OidcTenantConfig` in dynamic resolvers only

That is what I meant as well.

**Comment by sberyozkin** [comment]
What should we do though about users directly accessing `io.quarkus.oidc.OidcTenantConfig` to check some of its properties, as in the closed issue #44077, by injecting `TenantConfigBean` and checking some `io.quarkus.oidc.OidcTenantConfig` individual properties ?

I wonder if we should only keep `new  io.quarkus.oidc.OidcTenantConfig()` deprecated (the other constructor can be made private instead of deprecated I think, users don't use it), but keep all the `getters` non-deprecated (if some individual fields do not  have getters and users will get deprecated warnings when accessing fields then I can add temp getters) ?

As far as `for removal` is concerned, I can do it myself a bit later, not critical for this PR

**Comment by michalvavrik** [comment]
And about the getters, some return deprecated sub-classes that we could drop, so it is better if users use the new interface methods. Otherwise we need to copy & paste between this classes.

**Comment by sberyozkin** [comment]
@michalvavrik OK, thanks, please only hide the helper `OidcTenantConfig` constructor which was only added in 3.17 to facilitate the transition to config mappings, and there will be time to tune a few things before 3.18 is out by the end of Jan :-)

Thanks very much for taking care of this massive restructuring

**Review by sberyozkin — APPROVED** [review]

**Comment by michalvavrik** [comment]
> @michalvavrik OK, thanks, please only hide the helper OidcTenantConfig constructor which was only added in 3.17 to facilitate the transition to config mappings, and there will be time to tune a few things before 3.18 is out by the end of Jan :-)

done

> Thanks very much for taking care of this massive restructuring

yeah, this took a lot of patience, thanks for the review as well

**Pull request merged**
